### PR TITLE
busywork fleshing out crypto-system apis: key derivation, key exchange, and aead

### PR DESCRIPTION
**Pull request by neonphog**
## PR summary

- finishes off crypto system api!
- also marks old api deprecated

## Review checklist 
- [ ] The story has unit or integration tests
- [ ] No new bugs, and any tech-debt is identified and justified
- [ ] There is enough API documentation (how to use)
- [ ] There is enough code documentation (how the code works)

### CODE REVIEW

File: crates/crypto_api/src/crypto_system/mod.rs [hunk 1]
```diff
@@ -48,11 +54,23 @@ pub trait CryptoSystem: Sync {
...
+    fn kdf_max_bytes(&self) -> usize;
+
+    /// derive a new deterministic key based of index, context, and parent
+    fn kdf(
+        &self,
+        out_buffer: &mut Box<dyn Buffer>,
+        index: u64,
+        context: &Box<dyn Buffer>,
+        parent: &Box<dyn Buffer>,
+    ) -> CryptoResult<()>;
```
Comment (struktured) [inline crates/crypto_api/src/crypto_system/mod.rs]: It seems a little too easy to use the wrong buffer here in the argument list, maybe in a future iteration we switch to a builder pattern for this?

File: crates/crypto_api/src/crypto_system/mod.rs [hunk 2]
```diff
@@ -100,6 +118,85 @@ pub trait CryptoSystem: Sync {
...
+
+    fn aead_encrypt(
+        &self,
+        cipher: &mut Box<dyn Buffer>,
+        message: &Box<dyn Buffer>,
+        adata: Option<&Box<dyn Buffer>>,
+        nonce: &Box<dyn Buffer>,
+        secret: &Box<dyn Buffer>,
+    ) -> CryptoResult<()>;
+
```
Comment (struktured) [inline crates/crypto_api/src/crypto_system/mod.rs]: Same as above comment- builder pattern or something (wrapper types? a struct with fields equal to the existing function args?) to make this less error prone

File: crates/crypto_api/src/crypto_system/crypto_system_test.rs [hunk 3]
```diff
@@ -446,5 +656,222 @@ mod test {
...
+                return Err(CryptoError::BadNonceSize);
+            }
+
+            if secret.len() != self.aead_secret_bytes() {
+                return Err(CryptoError::BadSecretKeySize);
+            }
+
+            let cipher = cipher.read_lock();
+
+            if &cipher[2..4] != &nonce.read_lock()[..2] {
```
Comment (struktured) [inline crates/crypto_api/src/crypto_system/crypto_system_test.rs]: This code is cryptic, maybe deserves a comment about what is going here with the seemingly arbitrary array indexing.

File: crates/sodium/src/hash.rs [hunk 4]
```diff
@@ -14,6 +14,7 @@ pub const BYTES512: usize = rust_sodium_sys::crypto_hash_sha512_BYTES as usize;
 /// @param {SecBuf} input - the data to hash
 ///
 /// @param {SecBuf} output - Empty Buffer to be used as output
+#[deprecated(note = "use CryptoSystem")]
```
Comment (struktured) [inline crates/sodium/src/hash.rs]: What is purpose of this? To avoid ever breaking hc core? Are you going to audit hc core for deprecation warnings instead of build errors and make a separate PR in that repo?

Comment (neonphog) [reply crates/sodium/src/hash.rs]: exactly, core uses a bunch of these legacy apis still, it's not a trivial amount of work to convert them over to crypto system..

File: crates/crypto_api/src/crypto_system/crypto_system_test.rs (hunk 3 above)
Comment (neonphog) [reply crates/crypto_api/src/crypto_system/crypto_system_test.rs]: :+1:

File: crates/crypto_api/src/crypto_system/mod.rs (hunk 1 above)
Comment (neonphog) [reply crates/crypto_api/src/crypto_system/mod.rs]: named param macros!

File: crates/crypto_api/src/crypto_system/mod.rs (hunk 2 above)
Comment (neonphog) [reply crates/crypto_api/src/crypto_system/mod.rs]: named param macros!
